fuse_core/include/fuse_core/autodiff_local_parameterization.h

```diff
@@ -0,0 +1,222 @@
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_CORE_AUTODIFF_LOCAL_PARAMETERIZATION_H
+#define FUSE_CORE_AUTODIFF_LOCAL_PARAMETERIZATION_H
+
+#include <fuse_core/local_parameterization.h>
+#include <ceres/internal/autodiff.h>
+
+#include <memory>
+
+
+namespace fuse_core
+{
+
+/**
+ * @brief Create a local parameterization with the Jacobians computed via automatic differentiation.
+ *
+ * To get an auto differentiated local parameterization, you must define two classes with a templated operator()
+ * (a.k.a. a functor).
+ *
+ * The first functor should compute:
+ *
+ *   Plus(x, delta) -> x_plus_delta
+ *
+ * And the second functor should compute the inverse operation:
+ *
+ *   Minus(x1, x2) -> delta, such that Minus(x, Plus(x, delta)) = delta
+ *
+ * The autodiff framework substitutes appropriate "Jet" objects for the template parameter T in order to compute
+ * the derivative when necessary, but this is hidden, and you should write the function as if T were a scalar type
+ * (e.g. a double-precision floating point number).
+ *
+ * Additionally the GlobalSize and LocalSize must be specified as template parameters.
+ * - GlobalSize is the size of the variables x1 and x2. If this is a quaternion, the GloblaSize would be 4.
+ * - LocalSize is the size of delta, and may be different from GlobalSize. For quaternions, there are only 3 degrees
+ *   of freedom, so the LocalSize is 3.
+ *
+ * For more information on local parameterizations, see fuse_core::LocalParameterization
+ */
+template <typename PlusFunctor, typename MinusFunctor, int kGlobalSize, int kLocalSize>
```

Comment:
I assume the variable naming scheme is for consistency with Ceres.

Reply:
I was going to drop the leading `k`, but `ceres::LocalParameterization` defines `GlobalSize()` and `LocalSize()` methods. That means the template names cannot also be `GlobalSize` and `LocalSize`. Since I didn't have any better ideas, I went with the `kGlobalSize` and `kLocalSize` from the Ceres implementation. Happy to change it if you have a better suggestion.

Reply:
Nah, that's fine by me.